app/ts/context/BalanceContext.tsx

```diff
@@ -0,0 +1,94 @@
+import { Signal, useSignal } from '@preact/signals'
+import { BigNumber, ethers } from 'ethers'
+import { ComponentChildren, createContext } from 'preact'
+import { useContext } from 'preact/hooks'
+import { useAsyncState } from '../library/preact-utilities.js'
+import { assertsExternalProvider, assertUnreachable, isEthereumObservable } from '../library/utilities.js'
+
+type Balance =
+	| {
+			state: 'initial'
+			checkBalance: () => void
+	  }
+	| {
+			state: 'updating'
+			reset: () => void
+	  }
+	| {
+			state: 'updated'
+			balance: BigNumber
+			updated_at: Date
+	  }
+	| {
+			state: 'failed'
+			error: Error
+			reset: () => void
+	  }
+
+export type BalanceStore = Signal<Balance>
+export function createBalanceStore() {
+	const { value: query, waitFor, reset } = useAsyncState<BigNumber>()
+
+	const checkBalance = () => {
+		waitFor(async () => {
+			assertsExternalProvider(window.ethereum)
+			const provider = new ethers.providers.Web3Provider(window.ethereum)
+			const signer = provider.getSigner()
+			const address = await signer.getAddress()
+			return await provider.getBalance(address)
+		})
+	}
+
+	const balanceStoreDefaults = { state: 'initial' as const, checkBalance }
+	const balanceStore = useSignal<Balance>(balanceStoreDefaults)
+
+	const handleAccountChange = (newAccount: string[]) => {
+		if (ethers.utils.isAddress(newAccount[0])) {
+			checkBalance()
+		}
+	}
+
+	const listenForAccountChanges = () => {
+		assertsExternalProvider(window.ethereum)
+		const provider = new ethers.providers.Web3Provider(window.ethereum)
+		if (!isEthereumObservable(provider.provider)) return
+		provider.provider.on('accountsChanged', handleAccountChange)
```

Comment:
remember to remove this listener when its no longer needed

Reply:
The updated approach no longer relies on these events and instead directly referenced from the address established during connection, ensuring they are synced throughout changes.